recipes/upstream/google-cloud/gke/main.tf

```diff
@@ -1,47 +1,80 @@
-module "google-kubernetes-engine" {
-  source               = "../../../../modules/distribution/gke"
-  prefix               = var.prefix
-  project_id           = var.project_id
-  region               = var.region
-  kube_config_filename = var.kube_config_filename
-  #  vpc        = var.vpc
-  #  subnet     = var.subnet
+locals {
+  kc_path = var.kube_config_path != null ? var.kube_config_path : path.cwd
+  kc_file = var.kube_config_filename != null ? "${local.kc_path}/${var.kube_config_filename}" : "${local.kc_path}/${var.prefix}_kube_config.yml"
+}
+
+module "google_kubernetes_engine" {
+  source     = "../../../../modules/distribution/gke"
+  prefix     = var.prefix
+  project_id = var.project_id
+  region     = var.region
+  #  vpc                = var.vpc
+  #  subnet             = var.subnet
   #  cluster_version    = var.cluster_version
-  #  instance_count = var.instance_count
+  #  instance_count     = var.instance_count
   #  instance_disk_size = var.instance_disk_size
   #  disk_type          = var.disk_type
   #  image_type         = var.image_type
   #  instance_type      = var.instance_type
 }
 
-resource "helm_release" "ingress-nginx" {
-  name             = "ingress-nginx"
-  repository       = "https://kubernetes.github.io/ingress-nginx"
-  chart            = "ingress-nginx"
-  create_namespace = true
-  namespace        = "ingress-nginx"
+resource "null_resource" "first_setup" {
+  depends_on = [module.google_kubernetes_engine.kubernetes_cluster_node_pool]
+
+  provisioner "local-exec" {
+    command = "sh ${path.cwd}/first_setup.sh"
+  }
+}
+
+resource "local_file" "kube_config_yaml" {
+  depends_on = [null_resource.first_setup]
+
+  content = templatefile("../../../../modules/distribution/gke/kubeconfig.yml.tmpl", {
+    cluster_name    = module.google_kubernetes_engine.cluster_name,
+    endpoint        = module.google_kubernetes_engine.cluster_endpoint,
+    cluster_ca      = module.google_kubernetes_engine.cluster_ca_certificate,
+    client_cert     = module.google_kubernetes_engine.client_certificate,
+    client_cert_key = module.google_kubernetes_engine.client_key
+  })
+  file_permission = "0600"
+  filename        = local.kc_file
+}
+
+provider "kubernetes" {
+  config_path = local_file.kube_config_yaml.filename
+}
+
+provider "helm" {
+  kubernetes {
+    config_path = local_file.kube_config_yaml.filename
+  }
+}
 
-  set {
-    name  = "controller.service.type"
-    value = "LoadBalancer"
+resource "null_resource" "wait_ingress_services_startup" {
```

Comment:
Ideally, we'd avoid this as much as possible, I appreciate this might not be possible in this case.

How is the ingress controller being installed? With `helm_release` it might be possible to use the `wait` and `wait_jobs` attributes?

Reply:
There is also the `wait` variable for the rancher module
https://github.com/rancher/tf-rancher-up/blob/main/modules/rancher/variables.tf#L177-L180

Reply:
To avoid changing everything, I tried not to touch the Rancher module and/or the modules created for RKE, RKE2 and K3s (in the modules/distribution path).
The purpose of this review is to make all Google modules work well based on what exists in the repo... this could be a good idea for a new issue.

Reply:
Happy to be wrong but I don't think this would change other modules.. the thought was to try use another method instead of executing OS commands - i.e `command = "sleep ${var.waiting_time}"`

For eg:
- At the time of installing ingress-nginx, wait until deployment is complete (preferable), I couldn't see how ingress-nginx is installed so can't be specific.
- Use a [time_sleep](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/sleep) resource, or pass a value for `wait` to the rancher module (linked above, it also uses `time_sleep`)

Reply:
Initially of course I tried with nginx's helm_release resource, but for some reason there was a problem with the uninstall; it uninstalled the helm resource but did not delete some resources that kept the namespace in a terminating state...
To solve it, I added a series of local-exec resources that cleaned up, but the code seemed so heavy... We can open an issue to do all the necessary tests on rewriting the GKE recipe.

Regarding the wait resource,I was forced to add it because, trying to automatically create the LB on the cloud provider, sometimes it does it quickly and other times not (does it depend on the peaks in the region? Maybe that’s why...).
In any case, the wait there is useful to ensure that the LB is created.
Otherwise, I tried with local-exec of bash code which executes a loop of kubectl commands until the svc responds, but it seemed to make the code too heavy...